app/lib/search/index_simple.dart

```diff
@@ -378,22 +381,42 @@ class SimplePackageIndex implements PackageIndex {
   }
 
   _TextResults _searchText(Set<String> packages, String text) {
+    final sw = Stopwatch()..start();
     if (text != null && text.isNotEmpty) {
-      final List<String> words = splitForIndexing(text).toList();
-      final int wordCount = words.length;
+      final words = splitForIndexing(text).toSet().toList();
+      final wordCount = words.length;
+
+      // lookup longer words first, as it may restrict the result set better
+      words.sort((a, b) => -a.length.compareTo(b.length));
+
+      // limit the number of words to lookup
+      if (words.length > 20) {
+        words.removeRange(20, words.length);
+      }
+
       final pkgScores = <Score>[];
       final apiPagesScores = <Score>[];
+      bool aborted = false;
+
       for (String word in words) {
+        if (packages.isEmpty) break;
+        if (pkgScores.length > 5 && sw.elapsedMilliseconds > 500) {
+          aborted = true;
+          _logger.info(
+              'Aborted word lookup after ${pkgScores.length} words and ${sw.elapsedMilliseconds} ms.');
```

Comment:
Can we include some string we can use the make custom metrics for counting number of slow queryies, like `[pub-slow-search-query]` if we include we can easily make a custom metric...